auth/oidc_provider.go

```diff
@@ -0,0 +1,253 @@
+package auth
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coreos/pkg/timeutil"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/jonboulle/clockwork"
+)
+
+const (
+	MaxProviderConfigSyncInterval = 24 * time.Hour
+	MinProviderConfigSyncInterval = time.Minute
+)
+
+// ErrChecksumNilHttpResponse is the error returned by getChecksum if the input HTTP response is nil.
+var ErrChecksumNilHttpResponse = errors.New("can't calculate the checksum of nil response")
+
+// providerConfigSync represents the properties required for running provider metadata sync.
+type providerConfigSync struct {
+	config      *OIDCProvider   // config represents the OpenID Connect Provider configuration.
+	minInterval time.Duration   // minInterval represents the Minimum provider metadata sync interval.
+	maxInterval time.Duration   // maxInterval represents the Maximum provider metadata sync interval.
+	clock       clockwork.Clock // clock provides an interface for a clock.
+	checksum    string          // checksum represents the SHA1 checksum of provider metadata.
+	ExpiresAt   time.Time       // ExpiresAt represents the provider metadata expiry time.
+}
+
+// newProviderConfigSync returns a new providerConfigSync instance with the
+// clock, minInterval and maxInterval initialized.
+func newProviderConfigSync() *providerConfigSync {
+	return &providerConfigSync{
+		clock:       clockwork.NewRealClock(),
+		minInterval: MinProviderConfigSyncInterval,
+		maxInterval: MaxProviderConfigSyncInterval,
+	}
+}
+
+// Run runs the provider metadata sync task in background.
+func (s *providerConfigSync) Run() chan struct{} {
+	stop := make(chan struct{})
+	var next pcsStepper
+	next = &pcsStepNext{aft: time.Duration(0)}
+	go func() {
+		for {
+			select {
+			case <-s.clock.After(next.after()):
+				next = next.step(s.sync)
+			case <-stop:
+				return
+			}
+		}
+	}()
+	return stop
+}
+
+// sync polls the provider metadata from discovery endpoint and refresh the ID token verifier.
+func (s *providerConfigSync) sync() (time.Duration, error) {
+	discoveryURL := strings.TrimSuffix(s.config.Issuer, "/") + OIDCDiscoveryConfigPath
+	base.Debugf(base.KeyAuth, "Fetching custom provider config from %s", base.UD(discoveryURL))
+	req, err := http.NewRequest(http.MethodGet, discoveryURL, nil)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error building new request for URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error invoking calling discovery URL %s: %v", base.UD(discoveryURL), err)
+		return 0, err
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("unsuccessful response: %v", err)
+	}
+
+	var ttl time.Duration
+	var ok bool
+	ttl, ok, err = cacheable(resp.Header)
```

Comment:
Make sure that in cases that cacheable returns ttl=0, we're using the default interval (1h), not the min interval.

Reply:
Added changes to use the default interval of 24 hours when cachable returns ttl = 0.